### DISCUSSION TITLE
请教下消息队列的问题

### DISCUSSION THREAD

**Discussion by JavenLaw** [cloudwu/skynet#1924] [General]
skynet_mq.c文件：
```
// When the queue is create (always between service create and service init) ,
// set in_global flag to avoid push it to global queue .
// If the service init success, skynet_context_new will call skynet_mq_push to push it to global queue.

q->in_global = MQ_IN_GLOBAL;
```

即：先把 in_global 置为1，假装次级队列在全局消息队列中了；等服务初始化完成之后，再真正skynet_globalmq_push(queue)放入到全局消息队列

skynet_server.c文件：
```
if (ret) {
	ctx->init = true;
}
skynet_globalmq_push(queue);
```
此时真正把服务的消息放入全局消息队列

我记得几年前不知道哪里看过一篇文章，说了这样做的原因，现在再次看源码，就是想不起来

希望有知道的同学指点一下：

在服务刚创建的时候，为什么就需要把次级消息队列放到全局队列中去？

不能等有消息到来的时候，再通过 skynet_mq_push() 中的
```
if (q->in_global == 0) {
	q->in_global = MQ_IN_GLOBAL;
	skynet_globalmq_push(q);
}
```
自动放入全局消息队列呢？

**Comment by cloudwu** [comment]
依赖后面的流程有可能导致一个没完全初始化完毕的服务被消息调度器调度到。

**Comment by JavenLaw** [comment]
今天看了下service_snlua.c的代码

1，调用skynet_context_new()函数时，先一步建立了服务的消息队列
```
// skynet_start.c文件，154行
ctx->queue = skynet_mq_create(ctx->handle)
```

2，snlua_init()函数，给自己发了第一条信息，因此是能放进snlua的消息队列的
```
// service_snlua文件，476行
// it must be first message
skynet_send(ctx, 0, ctx->handle, PTYPE_TAG_DONTCOPY,0, tmp, sz);
```

3，流程总结：
skynet_context_new()创建服务 --> 建立snlua的消息队列  -->  snlua_init()函数开始  -->  给自己发第一条信息  -->  snlua消息队列收到信息   -->  snlua_init()函数结束    -->  把snlua消息队列放到全局队列

如果此时等有消息到来的时候，就立即把snlua消息队列放入全局队列
会导致snlua_init()函数没执行完毕，skynet_context_new()也没初始化完，就被worker线程获取执行

就是云大说的：依赖后面的流程有可能导致一个没完全初始化完毕的服务被消息调度器调度到

希望能帮助有需要的人，也算给自己做个笔记
